File: Algorithms/DepthFirstSearch/BinaryTree.py
from typing import *
from typing_extensions import *

class Node:
  def __init__(self:Self, val:int, left:Self|None=None, right:Self|None=None) -> None:
    self.val=val
    self.left=left
    self.right=right

  # DFS Algorithm
  def __str__(self):
    return f"{self.val}[{str(self.left) if self.left else '-'}|{str(self.right) if self.right else '-'}]"
    
def main() -> None:
  tree = Node(1,
    Node(2,
      Node(3),
      Node(4)
    ),
    Node(5,
      Node(6),
      Node(7)
    )
  )

  print(tree)

# Main
if __name__ == "__main__": main();

File: Algorithms/DepthFirstSearch/BinaryTree.cpp
#include <iostream>
#include <string>

class Node {
public:
  int val;
  Node* left;
  Node* right;

  Node(int val, Node* left = nullptr, Node* right = nullptr)
    : val(val), left(left), right(right) { }

  ~Node() {
    if (left) { delete left; }
    if (right) { delete right; }
  }

  std::string String() { 
    std::string str = "";
    str += std::to_string(val);
    str += "[";
    str += (left ? left->String() : "-");
    str += "|";
    str += (right ? right->String() : "-");
    str += "]";
    return str;
  }
};

int main() {

  Node* root = new Node(1,
    new Node(2,
      new Node(4),
      new Node(5)
    ),
    new Node(3,
      new Node(6),
      new Node(7)
    )
  );

  std::cout << root->String() << std::endl;

  delete root;

  return 0;
}